Add TaHoma integration (base + Binary Sensor)

iMicknl opened the pull request:
## Proposed change
Since #41267 is not easy to review, we (@vlebourl, @tetienne and me) will split the work done in https://github.com/imicknl/ha-tahoma in multiple small PR's. This PR includes some of the ground work for other platforms, but all non necessary code has been removed to keep it as small as possible.

- Add new Somfy TaHoma domain and integration, using Config Flow and DataUpdateCoordinator.
- Support for Binary Sensors, other platforms will be added in other PR's.

Would be great to PR this to a feature branch, instead of dev, since I would like to have all platforms (especially cover) included in a core release.

## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issues: 
	- fixes #37389
	- fixes #37291
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/14920

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.
	- [x] 37717
	- [x] 40626

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

frenck commented:
Why the new domain? If the old integration is broken, you can just replace it, right?

iMicknl commented:
> Why the new domain? If the old integration is broken, you can just replace it, right?

See discussion at https://github.com/home-assistant/core/pull/43952#issuecomment-739248051.

iMicknl commented:
> Why the new domain? If the old integration is broken, you can just replace it, right?

I am not sure what we all agreed on in the end in #43952. Should we change this PR to target `tahoma` instead of `somfy_tahoma`, or shall we take this opportunity to update the domain to a more suitable name?

Our reason for targeting `somfy_tahoma` is mainly because this integration is called Somfy TaHoma nowadays and it is closely related to the original `somfy` integration. It would be good to have them close in the code, like `somfy_mylink`, to ease the discovery for new developers. I don't consider it a breaking change since the current integration is not in use and broken for many months, but if the core team doesn't allow this change, we will change it back to `tahoma`.

frenck commented:
I would say; keep the domain as is. Another helpful thing for users; is migration. The username/password from an existing `tahoma` config can be imported as a config entry.

iMicknl commented:
> I would say; keep the domain as is. Another helpful thing for users; is migration. The username/password from an existing `tahoma` config can be imported as a config entry.

Done! This PR will now target the `tahoma` domain.

iMicknl committed:
Add refactored Somfy TaHoma integration

iMicknl commented:
Lately, I accidentally rebased on latest dev, but I forgot that the target is `rework-tahoma`. Force pushed a fixed version rebased on `rework-tahoma`.

If this PR is reviewed / merged to the work branch, I will add all other (small) PR's, however we first need to have the base merged for that.

MartinHjelmare commented on homeassistant/components/tahoma/config_flow.py:
```diff
@@ -0,0 +1,107 @@
+    async def async_step_update_interval(self, user_input=None):
+        """Manage the options regarding interval updates."""
+        if user_input is not None:
+            self.options[CONF_UPDATE_INTERVAL] = user_input[CONF_UPDATE_INTERVAL]
```
Not sure if this has any effect. A new options flow instance is created every time we start the options flow.

iMicknl replied:
Thanks, I will have a look and make the change if this code isn't required. I added it based on other integrations, but it could be that it isn't required :).

iMicknl replied:
Fixed!

MartinHjelmare commented on homeassistant/components/tahoma/coordinator.py:
```diff
@@ -0,0 +1,150 @@
+        self.devices: Dict[str, Device] = {d.deviceurl: d for d in devices}
+        self.executions: Dict[str, Dict[str, str]] = {}
+
+        _LOGGER.debug(
```
We don't want side effects like logging in init methods.

iMicknl replied:
@MartinHjelmare ok, I thought that extra logging, as long as its debug, is possible. What would be the best way to add debug logging (thus no info / warning). It is really helpful to understand someones update interval while investigating authentication issues. 

Do you know any other integrations that do this in the right way? :)

MartinHjelmare replied:
We can log the updated interval after creating the coordinator outside it, or after updating the interval.

MartinHjelmare commented on homeassistant/components/tahoma/coordinator.py:
```diff
@@ -0,0 +1,150 @@
+        try:
+            events = await self.client.fetch_events()
+        except BadCredentialsException as exception:
+            raise UpdateFailed("invalid_auth") from exception
```
Please write a statement in the exception message, not a single word.

iMicknl replied:
Fixed.

MartinHjelmare commented on homeassistant/components/tahoma/manifest.json:
```diff
@@ -0,0 +1,15 @@
+  "requirements": [
+    "pyhoma==0.5.1"
+  ],
+  "dependencies": [],
```
```suggestion
```

iMicknl replied:
Fixed.

MartinHjelmare commented on homeassistant/components/tahoma/tahoma_device.py:
```diff
@@ -0,0 +1,169 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class TahomaDevice(CoordinatorEntity, Entity):
```
```suggestion
class TahomaEntity(CoordinatorEntity, Entity):
```

iMicknl replied:
Fixed.

MartinHjelmare commented on homeassistant/components/tahoma/tahoma_device.py:
```diff
@@ -0,0 +1,169 @@
+    def device_state_attributes(self) -> Dict[str, Any]:
+        """Return the state attributes of the device."""
+        attr = {
+            "ui_class": self.device.ui_class,
```
Is this static info? We only want dynamic info related to the state of the entity in state attributes.

iMicknl replied:
Removed.

MartinHjelmare commented on homeassistant/components/tahoma/tahoma_device.py:
```diff
@@ -0,0 +1,169 @@
+        }
+
+        if self.has_state(CORE_RSSI_LEVEL_STATE):
+            attr[ATTR_RSSI_LEVEL] = self.select_state(CORE_RSSI_LEVEL_STATE)
```
This should be a separate sensor entity:
https://developers.home-assistant.io/docs/core/entity/sensor#available-device-classes

iMicknl replied:
Removed for now.

MartinHjelmare commented on homeassistant/components/tahoma/tahoma_device.py:
```diff
@@ -0,0 +1,169 @@
+            elif battery_state == STATE_BATTERY_VERY_LOW:
+                battery_state = 10
+
+            attr[ATTR_BATTERY_LEVEL] = battery_state
```
This should be a battery sensor entity.

iMicknl replied:
Is it okay to keep those attributes, until we have a way to separate those attributes to new entities. It is something that we have open for a while https://github.com/iMicknl/ha-tahoma/issues/223, but which is hard with all the different Somfy devices and characteristics. 

I would like to do this in a follow up PR. (especially since the target in this PR is not the dev repo yet).

MartinHjelmare replied:
For now, don't add the attributes at all. Follow up later and add the other sensors.

iMicknl replied:
Removed for now.

MartinHjelmare commented on homeassistant/components/tahoma/tahoma_device.py:
```diff
@@ -0,0 +1,169 @@
+
+        if self.device.attributes:
+            for attribute in self.device.attributes:
+                attr[attribute.name] = attribute.value
```
Please set the attribute keys explicitly so we can see what they are and their formatting. Eg by creating a constant dict that maps state attribute key to device attribute key, and iterating over that dict while looking up the device attribute values.

iMicknl replied:
All attributes removed for now, others could be added in a follow up PR.

MartinHjelmare commented on homeassistant/components/tahoma/tahoma_device.py:
```diff
@@ -0,0 +1,169 @@
+        """Return True if a state exists in self."""
+        return self.select_state(*states) is not None
+
+    async def async_execute_command(self, command_name: str, *args: Any):
```
I recommend moving all device api helpers to a separate class that represents the device and give the entity access to an instance of that class. Mixing in the device api in the entity api often leads to bad patterns.

iMicknl replied:
@tetienne, @vlebourl would you know how to do this? I thought about creating a `OverkizDevice` class, however in almost all device specific features we need the coordinator data / client. Not sure how this would work with inheritance in Python.

iMicknl replied:
Done!

iMicknl committed:
Rename TahomaDevice to TahomaEntity

iMicknl committed:
Remove custom icon

iMicknl committed:
Simplify attributes for core PR..

iMicknl committed:
Remove empty keys from manifest.json

iMicknl committed:
Change logging to use full sentences.

iMicknl commented on homeassistant/components/tahoma/__init__.py:
```diff
@@ -0,0 +1,139 @@
+        await client.login()
+        devices = await client.get_devices()
+    except BadCredentialsException:
+        _LOGGER.error("invalid_auth")
```
Fixed.

iMicknl committed:
Rename entities to devices

iMicknl commented on homeassistant/components/tahoma/__init__.py:
```diff
@@ -0,0 +1,139 @@
+
+    await tahoma_coordinator.async_refresh()
+
+    entities = defaultdict(list)
```
Changed to `devices`.

iMicknl commented on homeassistant/components/tahoma/binary_sensor.py:
```diff
@@ -0,0 +1,116 @@
+        ) or TAHOMA_BINARY_SENSOR_DEVICE_CLASSES.get(self.device.ui_class)
+
+    @property
+    def icon(self) -> Optional[str]:
```
Removed. What would be the best way to do this? I still see this behaviour in quite some integrations.

MartinHjelmare replied:
Just set device class. Entities with device class should already have an icon set in the frontend.

iMicknl replied:
We only set an icon for devices where no device class exists. (behavior from old `tahoma` integration). I will have a look and otherwise create an architecture issue/request.

iMicknl committed:
Bump pyhoma to latest version

iMicknl committed:
Fix device state bug

iMicknl committed:
Regenerate requirements

iMicknl committed:
Fix .coveragerc after recent changes

iMicknl commented:
@MartinHjelmare thanks for your review. Appreciated!

We will work on addressing your feedback. I see that `CI / Prepare tests for Python 3.8 (pull_request)` is failing, but it doesn't look related to this integration.

Could it be that it this branch + `home-assistant:rework-tahoma` are too old, or am I missing something else? If `home-assistant:rework-tahoma` could be rebased on dev, I am happy to rebase this branch on dev as well.

iMicknl committed:
Remove typing

MartinHjelmare commented:
Yeah, I'll rebase the target branch and then you can rebase on top of it.

iMicknl committed:
Remove device attributes for now

iMicknl committed:
Improve logging

iMicknl committed:
Fix OptionsFlow

iMicknl committed:
Simplify OptionsFlow

iMicknl committed:
Remove unused constants

iMicknl committed:
Rename devices to platforms (makes more sense)

MartinHjelmare commented:
I've rebased the target branch.

iMicknl committed:
Move non entity logic to seperate class

iMicknl committed:
Fix binary_sensor

iMicknl committed:
Fix arguments order

iMicknl commented:
Processed all feedback and all checks have passed. ✅

MartinHjelmare commented on .coveragerc:
```diff
@@ -918,6 +918,10 @@ omit =
+    homeassistant/components/tahoma/__init__.py
+    homeassistant/components/tahoma/binary_sensor.py
+    homeassistant/components/tahoma/coordinator.py
+    homeassistant/components/tahoma/tahoma_entity.py
```
Exclude the `overkiz_executor.py` module too.

iMicknl replied:
Fixed.

MartinHjelmare commented on homeassistant/components/tahoma/binary_sensor.py:
```diff
@@ -0,0 +1,97 @@
+
+    entities = [
+        TahomaBinarySensor(device.deviceurl, coordinator)
+        for device in data["platforms"].get(BINARY_SENSOR)
```
```suggestion
        for device in data["platforms"][BINARY_SENSOR]
```

iMicknl replied:
`data["platforms"][BINARY_SENSOR]` is optional, since not all Somfy TaHoma devices will require a binary_sensor. Would it be okay to keep the `get(BINARY_SENSOR)` in this scenario?

MartinHjelmare replied:
We only forward the entry to the platform if the platform is in `platforms`. Did I miss something?

iMicknl replied:
No, you are right. Oops! :D

iMicknl replied:
Fixed.

MartinHjelmare commented on homeassistant/components/tahoma/binary_sensor.py:
```diff
@@ -0,0 +1,97 @@
+CORE_VIBRATION_STATE = "core:VibrationState"
+CORE_WATER_DETECTION_STATE = "core:WaterDetectionState"
+
+DEVICE_CLASS_BUTTON = "button"
+DEVICE_CLASS_GAS = "gas"
+DEVICE_CLASS_RAIN = "rain"
+DEVICE_CLASS_WATER = "water"
+
```
Only standard entity device classes may be used. Import and use the constants from the binary sensor integration.

https://developers.home-assistant.io/docs/core/entity/binary-sensor#available-device-classes

```suggestion
```

iMicknl replied:
Will do. I should investigate if other devices could use `DEVICE_CLASS_RAIN` or `DEVICE_CLASS_WATER` as well, for a new architecture proposal.

iMicknl replied:
Fixed.

MartinHjelmare commented on homeassistant/components/tahoma/tahoma_entity.py:
```diff
@@ -0,0 +1,73 @@
+    @property
+    def assumed_state(self) -> bool:
+        """Return True if unable to access real state of the entity."""
+        return self.device.states is None or len(self.device.states) == 0
```
```suggestion
        return not self.device.states
```

iMicknl replied:
@tetienne, @vlebourl do you know where this is coming from? Is there an edge case where we need to check the length of `self.device.states`? I recall that we have situations where `self.device.states` is an empty array, which doesn't work with `not self.device.states`.

tetienne replied:
The suggestion William work too that's equivalent.
That's for RTS devices if I remembered well.

iMicknl replied:
I just tested it and indeed, `not` tackles our use-case as well. Good change :)

```python
states = {} # or []

if not states:
    print("No states")
else:
    print("States exists, but could be empty")
```

iMicknl replied:
Fixed.

iMicknl committed:
Add overkiz_executor.py to coveragerc

iMicknl committed:
Only use predefined device classes

iMicknl committed:
Bugfix

iMicknl committed:
Don’t use get to pull the dictionary

iMicknl committed:
Simplify assumed states

iMicknl commented:
Feedback of second round processed, hopefully all checks will pass again.

Minor question, after we removed `translations/en.json` for this PR, the dialog is not showing any strings. Is this a mistake on our side, or will this be fixed after the first sync with Lokalise?

<img width="405" alt="Screen Shot 2021-02-22 at 13 50 53" src="https://user-images.githubusercontent.com/1424596/108711105-2cb8be80-7515-11eb-8f19-ccb078fc788f.png">

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
Yeah, translations are needed to test the frontend, and this will be updated by our infrastructure and Lokalise. But we do allow committing the default English translation file for testing purposes and convenience.

iMicknl commented:
> Yeah, translations are needed to test the frontend, and this will be updated by our infrastructure and Lokalise. But we do allow committing the default English translation file for testing purposes and convenience.

Ok, I will add the English one to this PR. Thanks!

iMicknl committed:
Add English translation for convenience

bdraco reviewed (approved):
LGTM

bdraco commented on homeassistant/components/tahoma/tahoma_entity.py:
```diff
@@ -0,0 +1,73 @@
+            "manufacturer": manufacturer,
+            "name": self.name,
+            "model": model,
+            "sw_version": self.device.controllable_name,
```
I think this device can set `suggested_area` here in a future PR.

iMicknl replied:
We created it as a work item already on our custom component repo, https://github.com/iMicknl/ha-tahoma/issues/392 :). However, it would require pulling the TaHoma area's first and that endpoint hasn't been implemented yet in our client.

MartinHjelmare commented:
Please link a docs PR in the PR description. Make a note in the docs PR that there will be multiple parent PRs in the core repo before the docs PR should be merged. Then we can merge here.

iMicknl commented:
> Please link a docs PR in the PR description. Make a note in the docs PR that there will be multiple parent PRs in the core repo before the docs PR should be merged. Then we can merge here.

@MartinHjelmare is it ok to add the docs for the final PR to `dev`? This PR is currently targeting a work branch. Otherwise, I would need to do a docs PR per PR, which is quite some work. I expect 10+ follow up PR's, since we will split this up by platform and amount of code change.

MartinHjelmare commented:
Yeah, the idea is that we use a single docs PR for this work.
